algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+
+theorem maximal_ideal_one_add_unit_to_local {α : Type u} [comm_ring α]
+  (m : set α) [is_ideal α m] [is_maximal_ideal m] :
+  (∀ x:α, x ∈ m → is_unit (1 + x)) → is_local α :=
+λ h, have hm : m = nonunits α, from set.ext $ λ x,
+  ⟨λ hx, not_unit_of_mem_maximal_ideal m _inst_3 hx,
+    λ hx, or.cases_on (is_maximal_ideal.no_between (m + principal_ideal x) (is_ideal.subset_add_left))
+      (λ h1, ((set.set_eq_def _ _).1 h1 x).1 ⟨0, x, is_ideal.zero_mem m, mem_principal_ideal x, by simp⟩)
+      (λ h1, let ⟨y, z, hy, ⟨p, hz⟩, hyz⟩ := ((set.set_eq_def _ _).1 h1 1).2 trivial in
+       let ⟨q, hq⟩ := h (-y) (is_ideal.neg_mem hy) in
+       false.elim $ hx ⟨p * q, calc
+         x * (p * q) = (1 + -y) * q : by rw [hyz, ← hz]; ring
+                 ... = 1 : hq⟩)⟩,
+by rw hm at _inst_2; exact nonunits_ideal_to_local _inst_2
+
+-- Proposition 1.6 end
+
+
+class is_pid (α : Type u) [comm_ring α] extends is_integral_domain α : Prop :=
+(h : ∀ S [is_ideal α S], ∃ x, S = principal_ideal x)
+
+instance maximal_of_pid_of_prime_of_nonzero (α : Type u) [comm_ring α] [is_pid α]
+(S : set α) [is_ideal α S] [is_prime_ideal S] : S ≠ zero_ideal α → is_maximal_ideal S :=
+begin
+  intro h,
+  constructor,
+  apply is_prime_ideal.ne_univ_ideal,
+  intros T _ hst,
+  cases is_pid.h S with s hs,
+  cases is_pid.h T with t ht,
+  rw [hs,ht] at *,
+  cases hst (mem_principal_ideal s) with z hz,
+  have h1 : t * z ∈ S := by rw [hs,hz]; exact mem_principal_ideal s,
+  cases is_prime_ideal.mem_or_mem_of_mul_mem h1 with p p,
+  left,
+    apply set.eq_of_subset_of_subset _ hst,
+    apply principal_ideal.subset_of_mem,
+    rwa hs at p,
+  right,
+    rw [← principal_ideal_one_eq_univ],
+    rw [← unit_iff_principal_ideal_eq_one],
+    rw hs at p,
+    cases p with n hn,
+    have : z * (t * n - 1) = 0 := calc
+      z * (t * n - 1) = s * n - z : by rw ← hz; ring
+                  ... = 0 : by simp [hn],
+    cases is_integral_domain.zero_or_zero_of_mul_zero z (t * n - 1) this with p p,
+    simp [p] at hz,
+    exfalso,
+    apply h,
+    rw [← hz],
+    apply principal_ideal_zero_eq_zero_ideal,
+    rw sub_eq_zero at p,
+    exact ⟨n, p⟩
+end
+
+@[simp] lemma pow_coset {α : Type u} [comm_ring α] {S : set α} [is_ideal α S] (x : α) :
+Π (n : nat), ⟦x⟧^n = ⟦x^n⟧
+| 0            := rfl
+| (nat.succ n) := by unfold monoid.pow; simp [pow_coset n]
+
+-- for lack of a [better] name (or a name at all)
+def nil_aux_1 {α : Type u} [comm_ring α] (x y : α) : nat → α
+| 0            := 0
+| (nat.succ n) := (x + y)^n + nil_aux_1 n * y
+
+theorem nil_aux_2 {α : Type u} [comm_ring α] (x y : α) : Π n:nat, (x + y)^n - y^n = x * nil_aux_1 x y n
+| 0            := by simp [nil_aux_1]
+| (nat.succ n) := calc (x + y)^nat.succ n - y^nat.succ n
+      = x * (x + y)^n + y * (x * nil_aux_1 x y n) : by unfold monoid.pow; rw ← nil_aux_2 n; ring
+  ... = x * nil_aux_1 x y (nat.succ n)            : by unfold nil_aux_1; ring
+
```

Comment:
If we had the binomial theorem this might all be unnecessary

Reply:
I did prove the binomial theorem today. I need to tidy stuff up a bit before making a PR.